src/nvim/ui.c

```diff
@@ -182,6 +183,37 @@ static void ui_refresh_event(void **argv)
   ui_refresh();
 }
 
+void ui_notify_changed_highlights(garray_T changed_highlights)
+{
+  // we might want to send batchs instead
+  // + save current hl_id value to restore it later
+  // for backwards compatibility
+  ILOG("%d changed highlights", changed_highlights.ga_len);
+  Array args = ARRAY_DICT_INIT;
+  Error err;
+  bool refresh_cursor = false;
+  for (int idx = 0; idx < changed_highlights.ga_len; idx++) {
+    int hl_id = ((int *)changed_highlights.ga_data)[idx];
+    Dictionary dic = nvim_hl_from_id(hl_id, &err);
+    ADD(args, DICTIONARY_OBJ(dic));
+    // ILOG("comparing with %d", shape_table[cursor_get_mode_idx()].id);
+    if (hl_id == shape_table[cursor_get_mode_idx()].id) {
+      refresh_cursor = true;
+    }
+  }
+
+  if (args.size) {
+    ILOG("submit %d updated hl", args.size);
+    ui_event("highlights", (args));
```

Comment:
ui events aren't varadic, this should be a nested array. But changing to `ui_call_highlights_changed(args)` will take care of this.